plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,102 +67,135 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
-            }
             clusteringColumnSql.add(predicateString);
-            domainsBuilder.put(columnHandle, domain);
             // Check for last clustering column should only be restricted by range condition
             if (predicateString.contains(">") || predicateString.contains("<")) {
                 break;
             }
-            currentClusteringColumn++;
+            allProcesedClusteringColumns++;
         }
         List<String> clusteringColumnPredicates = clusteringColumnSql.build();
 
-        return new ClusteringPushDownResult(domainsBuilder.build(), Joiner.on(" AND ").join(clusteringColumnPredicates));
+        return new ClusteringPushDownResult(fullyPushedDomains.build(), Joiner.on(" AND ").join(clusteringColumnPredicates));
+    }
+
+    /**
+     * IN restriction allowed only on last clustering column for Cassandra version <= 2.1
+     */
+    private static boolean isInStatementNotAllowed(List<CassandraColumnHandle> clusteringColumns, VersionNumber cassandraVersion, int currentlyProcessedClusteringColumn)
+    {
+        return cassandraVersion.compareTo(VersionNumber.parse("2.2.0")) < 0 && currentlyProcessedClusteringColumn != (clusteringColumns.size() - 1);
+    }
+
+    private static String toCqlLiteral(CassandraColumnHandle columnHandle, Object value)
+    {
+        return columnHandle.getCassandraType().toCqlLiteral(value);
+    }
+
+    private static String translateRangeIntoCql(CassandraColumnHandle columnHandle, Range range)
+    {
+        if (range.isAll()) {
+            return null;
+        }
+        if (range.isSingleValue()) {
+            return format("%s = %s",
+                    CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                    toCqlLiteral(columnHandle, range.getSingleValue()));
+        }
+
+        String lowerBoundPredicate = null;
+        String upperBoundPredicate = null;
+        if (!range.isLowUnbounded()) {
+            String lowBound = toCqlLiteral(columnHandle, range.getLowBoundedValue());
+            lowerBoundPredicate = format(
+                    "%s %s %s",
+                    CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                    range.isLowInclusive() ? ">=" : ">",
+                    lowBound);
+        }
+        if (!range.isHighUnbounded()) {
+            String highBound = toCqlLiteral(columnHandle, range.getHighBoundedValue());
+            upperBoundPredicate = format(
+                    "%s %s %s",
+                    CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                    range.isHighInclusive() ? "<=" : "<",
+                    highBound);
+        }
+        if (lowerBoundPredicate != null && upperBoundPredicate != null) {
+            return format("%s AND %s ", lowerBoundPredicate, upperBoundPredicate);
+        }
+        if (lowerBoundPredicate != null) {
+            return lowerBoundPredicate;
+        }
+        return upperBoundPredicate;
```

Comment:
Can ranges be "empty" (looking at the Range class makes it look possible)? If so this can return `null`. Is this expected?